dtfs-central-api/src/helpers/get-fee-record-payment-entity-group-reconciliation-data.ts

```diff
@@ -0,0 +1,59 @@
+import { orderBy } from 'lodash';
+import { FEE_RECORD_STATUS } from '@ukef/dtfs2-common';
+import { TfmUsersRepo } from '../repositories/tfm-users-repo';
+import { FeeRecordReconciledByUser } from '../types/utilisation-reports';
+import { FeeRecordPaymentEntityGroup } from '../types/fee-record-payment-entity-group';
+
+/**
+ * Gets the reconciliation data for a fee record payment entity group
+ * @param group - The fee record payment entity group
+ * @returns The group reconciliation data
+ * @throws {Error} If the groups fee records array is empty
+ * @throws {Error} If any of the fee records at the `RECONCILED` status have a null `dateReconciled` property
+ */
+export const getFeeRecordPaymentEntityGroupReconciliationData = async (
+  group: FeeRecordPaymentEntityGroup,
+): Promise<{
+  reconciledByUser?: FeeRecordReconciledByUser;
+  dateReconciled?: Date;
+}> => {
+  if (group.feeRecords.length === 0) {
+    throw new Error('Fee record payment entity group cannot have an empty fee records array');
+  }
+
+  const reconciledFeeRecords = group.feeRecords.filter(({ status }) => status === FEE_RECORD_STATUS.RECONCILED);
+
+  if (reconciledFeeRecords.length === 0) {
+    return {};
+  }
+
+  if (reconciledFeeRecords.some(({ dateReconciled }) => dateReconciled === null)) {
+    throw new Error(`Fee records at the '${FEE_RECORD_STATUS.RECONCILED}' status cannot have a null 'dateReconciled' property`);
+  }
+
+  const feeRecordsSortedByDateReconciledDescending = orderBy(reconciledFeeRecords, [(feeRecord) => feeRecord.dateReconciled!.getTime()], ['desc']);
+  const mostRecentlyReconciledFeeRecord = feeRecordsSortedByDateReconciledDescending.at(0);
+  if (!mostRecentlyReconciledFeeRecord) {
+    // Based on all the above checks, this error should never be reached, hence not being documented
+    throw new Error('Something went wrong when sorting the fee records to get the most recently reconciled fee record');
+  }
+
+  const { dateReconciled, reconciledByUserId } = mostRecentlyReconciledFeeRecord;
+
+  if (!reconciledByUserId) {
+    return { dateReconciled: dateReconciled! };
+  }
```

Comment:
Should we be returning something here if there is no user ID associated?

Reply:
I don't think this is ever expected to happen, but if it did happen we would want the page to still display rather than return problem with service and so it makes sense I think to return the data we do have.